academy/golang-intro/7_concurrency.md

```diff
@@ -0,0 +1,221 @@
+---
+title: "Concurrency in Go"
+order: 
+description: 
+tag: deep-dive
+---
+
+# Concurrency in Go
+
+Go has built-in concurrency. Note, that concurrency and parallel execution are different things; you need concurrency (*program structure*) to enable parallel execution. Actual parallelism during execution depends on hardware. 
+
+For concurrency, Go offers so called *Goroutines* and *Channels*.
+
+## Goroutines
+
+A *Gouroutine* is a concurrent thread managed by the Go runtime. 
+
+Let's see, how to call a goroutine:
+
+```golang
+package main
+
+import (
+    "fmt"
+    "time"
+)
+
+func doSomething(size int) {
+    for i := 0; i < size; i++ {
+        fmt.Println(i)
+        time.Sleep(time.Second)
+    }
+}
+
+func main() {
+    go doSomething(10) // go statement before a function creates a goroutine
+    go doSomething(10)
+    time.Sleep(10*time.Second)
+}
+
+```
+
+<HighlightBox type="tip">
+
+[Test it online](https://go.dev/play/p/6c1vJ2Xz9WB).
+
+</HighlightBox>
+
+If you run this programm, you will see, that both `doSomething(10)` functions work concurrently. We wait with `time.Sleep(10*time.Second)` to see this in action. 
+
+## Channels
+
+Go offers *channels* for commincation between goroutines. Channels may be buffered or unbuffered. First, let's see how we can create an *unbuffered* channel:
+
+```golang
+ch:= make(chan type)
+```
+
+Now, we can use this channel to send and receive messages. We use the `<-` operator:
+
+```golang
+ch <- v
+```
+
+to send to channel `ch` and:
+
+```golang
+v := <-ch
+```
+
+to read from channel `ch`.
+
+Ok let us write an example using channels:
+
+```golang
+package main
+
+import (
+    "fmt"
+    "time"
+)
+
+func doSomething(size int, c chan int) {
+    for i := 0; i < size; i++ {
+        time.Sleep(100 * time.Millisecond)
+    }
+    c <- size
+}
+
+func main() {
+    c := make(chan int)
+    go doSomething(10, c)
+    go doSomething(20, c)
+    go doSomething(30, c)
+    
+    x, y, z := <-c, <-c, <-c
+    fmt.Println(x, y, z)
+}
+```
+
+<HighlightBox type="tip">
+
+[Test it online](https://go.dev/play/p/MYdZRhyG36y)
+
+</HighlightBox>
+
+Run this programm. What happened there?
+
+We don't need to use `time.Sleep` anymore because sends and receives block until the other side is ready. To avoid blocking, one can create *buffered* channels.
```

Comment:
In **109**, I am reading the word "block" as if it is being used in the sense of "prevent something happening" (same with "blocking"). 

However, in **115**, we have the following: _"When a *buffered* channel is full, **sends block**. When and if it is empty the channel **receives blocks**."_ This use of "blocks" reads more like the idea of _blockchain_ blocks, as in "a collection of data which is sent or received".

Are we using two different senses of the word "block" in close proximity here? If not, which is the sense both lines should be understood as?

Reply:
So, for 115:

When a *buffered* channel is full, sends to it are blocked. When a buffered channel is empty, receives from it are blocked.

Does that read appropriately?

Reply:
Perfect!